presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/MergingHashAggregationBuilder.java

```diff
@@ -0,0 +1,134 @@
+
+import java.io.Closeable;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+public class MergingHashAggregationBuilder
+    implements Closeable
+{
+    private final List<AccumulatorFactory> accumulatorFactories;
+    private final AggregationNode.Step step;
+    private final int expectedGroups;
+    private final ImmutableList<Integer> groupByPartialChannels;
+    private final Optional<Integer> hashChannel;
+    private final OperatorContext operatorContext;
+    private final Iterator<Page> sortedPages;
+    private InMemoryHashAggregationBuilder hashAggregationBuilder;
+    private final List<Type> groupByTypes;
+    private final LocalMemoryContext systemMemoryContext;
+    private final long memorySizeBeforeSpill;
+    private final int overwriteIntermediateChannelOffset;
+
+    public MergingHashAggregationBuilder(
+            List<AccumulatorFactory> accumulatorFactories,
+            AggregationNode.Step step,
+            int expectedGroups,
+            List<Type> groupByTypes,
+            Optional<Integer> hashChannel,
+            OperatorContext operatorContext,
+            Iterator<Page> sortedPages,
+            LocalMemoryContext systemMemoryContext,
+            long memorySizeBeforeSpill,
+            int overwriteIntermediateChannelOffset)
+    {
+        ImmutableList.Builder<Integer> groupByPartialChannels = ImmutableList.builder();
+        for (int i = 0; i < groupByTypes.size(); i++) {
+            groupByPartialChannels.add(i);
+        }
+
+        this.accumulatorFactories = accumulatorFactories;
+        this.step = AggregationNode.Step.partialInput(step);
+        this.expectedGroups = expectedGroups;
+        this.groupByPartialChannels = groupByPartialChannels.build();
+        this.hashChannel = hashChannel.isPresent() ? Optional.of(groupByTypes.size()) : hashChannel;
+        this.operatorContext = operatorContext;
+        this.sortedPages = sortedPages;
+        this.groupByTypes = groupByTypes;
+        this.systemMemoryContext = systemMemoryContext;
+        this.memorySizeBeforeSpill = memorySizeBeforeSpill;
+        this.overwriteIntermediateChannelOffset = overwriteIntermediateChannelOffset;
+
+        rebuildHashAggregationBuilder();
+    }
+
+    public Iterator<Page> buildResult()
+    {
+        return new Iterator<Page>() {
+            private Iterator<Page> resultPages = Collections.emptyIterator();
+
+            @Override
+            public boolean hasNext()
+            {
+                return sortedPages.hasNext() || resultPages.hasNext();
+            }
+
+            @Override
+            public Page next()
+            {
+                if (!resultPages.hasNext()) {
+                    rebuildHashAggregationBuilder();
+                    long memorySize = 0; // ensure that at least one merged page will be processed
+
+                    while (sortedPages.hasNext() && !shouldProduceOutput(memorySize)) {
```

Comment:
There was such comment over `PageRewriteIterator` in `MergeHashSort`, but I have moved to `merge` method.